File: 0501-0600/0518-coin-change-2.py
from typing import List

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1
        for x in coins:
            for i in range(x, amount + 1):
                dp[i] += dp[i - x]
        return dp[amount]

print(Solution().change(5, [1,2,5]))

File: 0501-0600/0518-coin-change-2.rs
struct Solution {}

impl Solution {
    pub fn change(amount: i32, coins: Vec<i32>) -> i32 {
        let mut dp = vec![0; (amount + 1) as usize];
        dp[0] = 1;
        for x in coins {
            for i in x..=amount {
                dp[i as usize] += dp[(i - x) as usize]
            }
        }
        dp[amount as usize]
    }
}

fn main() {
    assert_eq!(Solution::change(5, vec![1,2,5]), 4);
}
